framework/Models/EnsembleModel.py

```diff
@@ -438,35 +445,42 @@ def collectOutput(self,finishedJob,output):
+          inputsValuesOptionalOut    = inputsValuesOptionalOut if optionalOut.type != 'HistorySet' else inputsValuesOptionalOut.values()[-1]
           outputsValuesOptionalOut   = optionalOut.getParametersValues('outputs', nodeId = 'RecontructEnding')
+          outputsValuesOptionalOut   = outputsValuesOptionalOut if optionalOut.type != 'HistorySet' else outputsValuesOptionalOut.values()[-1]
           metadataValuesOptionalOut  = optionalOut.getAllMetadata(nodeId = 'RecontructEnding')
-          inputsValuesOptionalOut    = inputsValuesOptionalOut if optionalOut.type != 'HistorySet' else inputsValuesOptionalOut.values()[-1]
-          for key in self.modelsDictionary[modelIn]['OutputObject'][index].getParaKeys('inputs'):#  inputsValuesOptionalOut.items():
+          for key in self.modelsDictionary[modelIn]['OutputObject'][index].getParaKeys('inputs'):
             if key in inputsValuesOptionalOut:
               self.modelsDictionary[modelIn]['OutputObject'][index].updateInputValue (key,inputsValuesOptionalOut[key])
             else:
               self.raiseAWarning('Input key "'+key+'" is not contained in output '+self.modelsDictionary[modelIn]['OutputObject'][index].name)
-          for key in self.modelsDictionary[modelIn]['OutputObject'][index].getParaKeys('outputs'):#  inputsValuesOptionalOut.items():
+          for key in self.modelsDictionary[modelIn]['OutputObject'][index].getParaKeys('outputs'):
             if key in outputsValuesOptionalOut:
               self.modelsDictionary[modelIn]['OutputObject'][index].updateOutputValue (key,outputsValuesOptionalOut[key])
             else:
               self.raiseAWarning('Output key "'+key+'" is not contained in output '+self.modelsDictionary[modelIn]['OutputObject'][index].name)
           for key,value in metadataValuesOptionalOut.items():
             self.modelsDictionary[modelIn]['OutputObject'][index].updateMetadata(key,value[-1])
 
+    # collect the output of the STEP
     if output.type == 'HDF5':
       output.addGroupDataObjects({'group':self.name+str(finishedJob.identifier)},exportDict,False)
     else:
-      if output.name in exportDictTargetEvaluation.keys():
-        exportDict = exportDictTargetEvaluation[output.name]
-      for key in exportDict['inputSpaceParams' ] :
-        if key in output.getParaKeys('inputs'):
-          output.updateInputValue (key,exportDict['inputSpaceParams' ][key])
-      for key in exportDict['outputSpaceParams'] :
-        if key in output.getParaKeys('outputs'):
-          output.updateOutputValue(key,exportDict['outputSpaceParams'][key])
-      for key in exportDict['metadata']:
-        output.updateMetadata(key,exportDict['metadata'][key][-1])
+      optionalOutputNames = []
+      for key, value in optionalOutputs.items():
+        if key != 'uncollectedJobIds':
+          optionalOutputNames+=[obj.name for obj in value]
+      if output.name not in optionalOutputNames:
```

Comment:
No, because the presence of the outputs in the step is already checked in the initialization stage. This if is needed to skip the collecting of the outputs that are optional since it is performed before.